core/trino-main/src/main/java/io/trino/operator/aggregation/arrayagg/FlatArrayBuilder.java

```diff
@@ -0,0 +1,277 @@
+            Throwables.throwIfUnchecked(throwable);
+            throw new RuntimeException(throwable);
+        }
+    }
+
+    private void growCapacity()
+    {
+        if (openRecordGroup != null) {
+            closedRecordGroups.add(openRecordGroup);
+        }
+        openRecordGroup = new byte[recordSize * RECORDS_PER_GROUP];
+        capacity += RECORDS_PER_GROUP;
+    }
+
+    public void writeAll(BlockBuilder blockBuilder)
+    {
+        for (byte[] records : closedRecordGroups) {
+            int recordOffset = 0;
+            for (int recordIndex = 0; recordIndex < RECORDS_PER_GROUP; recordIndex++) {
+                write(records, recordOffset, blockBuilder);
+                recordOffset += recordSize;
+            }
+        }
+        int recordsInOpenGroup = ((int) size) & RECORDS_PER_GROUP_MASK;
+        int recordOffset = 0;
+        for (int recordIndex = 0; recordIndex < recordsInOpenGroup; recordIndex++) {
+            write(openRecordGroup, recordOffset, blockBuilder);
+            recordOffset += recordSize;
+        }
+    }
+
+    public long write(long index, BlockBuilder blockBuilder)
+    {
+        checkIndex(index, size);
+
+        byte[] records = getRecords(index);
+
+        int recordOffset = getRecordOffset(index);
+        write(records, recordOffset, blockBuilder);
+
+        if (hasNextIndex) {
+            return (long) LONG_HANDLE.get(records, recordOffset + recordNextIndexOffset);
+        }
+        return -1;
+    }
+
+    private void write(byte[] records, int recordOffset, BlockBuilder blockBuilder)
+    {
+        if (records[recordOffset + recordNullOffset] != 0) {
+            blockBuilder.appendNull();
+            return;
+        }
+
+        byte[] variableWidthChunk = EMPTY_CHUNK;
+        if (variableWidthData != null) {
+            variableWidthChunk = variableWidthData.getChunk(records, recordOffset);
+        }
+
+        try {
+            readFlat.invokeExact(
+                    records,
+                    recordOffset + recordValueOffset,
+                    variableWidthChunk,
+                    blockBuilder);
+        }
+        catch (Throwable throwable) {
+            Throwables.throwIfUnchecked(throwable);
+            throw new RuntimeException(throwable);
+        }
+    }
+
+    public FlatArrayBuilder copy()
+    {
+        return new FlatArrayBuilder(this);
+    }
+
+    private byte[] getRecords(long index)
+    {
+        int recordGroupIndex = (int) (index >>> RECORDS_PER_GROUP_SHIFT);
+        byte[] records;
+        if (recordGroupIndex < closedRecordGroups.size()) {
+            records = closedRecordGroups.get(recordGroupIndex);
+        }
+        else {
+            checkState(recordGroupIndex == closedRecordGroups.size());
+            records = openRecordGroup;
+        }
+        return records;
+    }
+
+    private int getRecordOffset(long index)
+    {
+        return (((int) index) & RECORDS_PER_GROUP_MASK) * recordSize;
```

Comment:
Why allow index to be `long` if this expects it to be an `int`? Would it make sense to change it to `int` everywhere?